File: reference-implementations/typescript/src/extract.ts
import { writeFile, readFile } from "node:fs/promises";
import { TomlDate } from "smol-toml";
import { CURRENT_TOML_SCHEMA_VERSION, type SchemaType } from "./builtins.js";
import { parseToml } from "./document.js";
import { encodeTomlKey } from "./paths.js";
import { isTomlArray, isTomlTable, type TomlTable, type TomlValue } from "./values.js";

/** Infers the built-in TOML Schema type name that best describes a parsed TOML value. */
function schemaTypeOf(value: TomlValue): SchemaType {
  if (typeof value === "string") return "string";
  if (typeof value === "bigint") return "integer";
  if (typeof value === "number") return "float";
  if (typeof value === "boolean") return "boolean";
  if (value instanceof TomlDate) {
    if (value.isDateTime()) return value.isLocal() ? "local-date-time" : "offset-date-time";
    if (value.isDate()) return "local-date";
    if (value.isTime()) return "local-time";
    return "any";
  }
  if (isTomlArray(value)) return "array";
  if (isTomlTable(value)) return "table";
  return "any";
}

/** Infers a homogeneous `itemtype` for an array, falling back to `"any"` when mixed or empty. */
function inferItemType(items: readonly TomlValue[]): SchemaType {
  if (items.length === 0) return "any";
  const first = schemaTypeOf(items[0] as TomlValue);
  for (const item of items.slice(1)) {
    if (schemaTypeOf(item) !== first) return "any";
  }
  return first;
}

function appendDefinition(lines: string[], path: readonly string[], value: TomlValue): void {
  lines.push("");
  lines.push(`[${path.map(encodeTomlKey).join(".")}]`);
  const typeName = schemaTypeOf(value);
  lines.push(`type = ${JSON.stringify(typeName)}`);
  if (typeName === "array" && isTomlArray(value)) {
    lines.push(`itemtype = ${JSON.stringify(inferItemType(value))}`);
  }
  if (isTomlTable(value)) {
    for (const childKey of Object.keys(value).sort()) {
      appendDefinition(lines, [...path, childKey], value[childKey] as TomlValue);
    }
  }
}

/**
 * Generates draft TOML Schema source text describing the structural shape of
 * a parsed TOML document. Every element is typed by inferring its TOML kind;
 * arrays additionally receive an inferred `itemtype` when homogeneous. The
 * root `[toml-schema]` table (if present in the document) is not itself
 * described, per the reserved-metadata convention.
 */
export function generateSchema(document: TomlTable): string {
  const lines: string[] = ["[toml-schema]", `version = ${JSON.stringify(CURRENT_TOML_SCHEMA_VERSION)}`, "", "[elements]"];
  for (const key of Object.keys(document).sort()) {
    if (key === "toml-schema") continue;
    appendDefinition(lines, ["elements", key], document[key] as TomlValue);
  }
  return lines.join("\n") + "\n";
}

/** Reads a TOML document, infers a draft schema for it, and writes the schema to `schemaPath`. */
export async function extractSchemaFile(documentPath: string, schemaPath: string): Promise<void> {
  const document = parseToml(await readFile(documentPath, "utf-8"));
  await writeFile(schemaPath, generateSchema(document), "utf-8");
}

File: reference-implementations/python/src/toml_schema/_extract.py
"""Starter-schema extraction from parsed TOML documents."""

from __future__ import annotations

import datetime
import json
import re
from pathlib import Path
from typing import Any

from ._schema import CURRENT_TOML_SCHEMA_VERSION, load_document

_BARE_KEY = re.compile(r"^[A-Za-z0-9_-]+$")


def generate_schema(document: dict[str, Any]) -> str:
    """Generates a draft TOML Schema describing a parsed TOML document."""
    lines = [
        "[toml-schema]",
        f'version = "{CURRENT_TOML_SCHEMA_VERSION}"',
        "",
        "[elements]",
    ]
    for key in sorted(document):
        if key != "toml-schema":
            _append_definition(lines, ("elements", key), document[key])
    return "\n".join(lines) + "\n"


def extract_schema_file(document_path: str, schema_path: str) -> None:
    """Reads a TOML document and writes its inferred draft schema."""
    schema = generate_schema(load_document(document_path))
    Path(schema_path).write_text(schema, encoding="utf-8")


def _append_definition(lines: list[str], path: tuple[str, ...], value: Any) -> None:
    lines.extend(("", f"[{'.'.join(_encode_key(part) for part in path)}]"))
    type_name = _schema_type(value)
    lines.append(f'type = "{type_name}"')
    if type_name == "array":
        lines.append(f'itemtype = "{_infer_item_type(value)}"')
    if isinstance(value, dict):
        for key in sorted(value):
            _append_definition(lines, path + (key,), value[key])


def _schema_type(value: Any) -> str:
    if isinstance(value, str):
        return "string"
    if isinstance(value, bool):
        return "boolean"
    if isinstance(value, int):
        return "integer"
    if isinstance(value, float):
        return "float"
    if isinstance(value, datetime.datetime):
        return "offset-date-time" if value.utcoffset() is not None else "local-date-time"
    if isinstance(value, datetime.date):
        return "local-date"
    if isinstance(value, datetime.time):
        return "local-time"
    if isinstance(value, list):
        return "array"
    if isinstance(value, dict):
        return "table"
    return "any"


def _infer_item_type(items: list[Any]) -> str:
    if not items:
        return "any"
    first = _schema_type(items[0])
    return first if all(_schema_type(item) == first for item in items[1:]) else "any"


def _encode_key(key: str) -> str:
    if _BARE_KEY.fullmatch(key):
        return key
    return json.dumps(key, ensure_ascii=False)
